### PR TITLE
Make CommandLineArguments fully typed

### PR DESCRIPTION
**Pull request by DaanDeMeyer**
This commit extends CommandLineArguments with all necessary fields
from the argument parser, removes the inheritance from argparse.Namespace
and fixes all resulting typing errors.

We try to stick to typing only changes as much as possible to reduce the
chance of breaking something (although there are a few non-typing changes
where doing so made things easier).

Because we use a dataclass now for the CommandLineArguments class, we
up the required python version to 3.7.

### CODE REVIEW

File: mkosi/__init__.py [hunk 1]
```diff
@@ -5577,7 +5738,7 @@ def print_summary(args: CommandLineArguments) -> None:
 
     MkosiPrinter.info("                 Read-only: " + yes_no(args.read_only))
     detail = " ({})".format(args.compress) if args.compress and not isinstance(args.compress, bool) else ""
-    MkosiPrinter.info("            FS Compression: " + yes_no(args.compress) + detail)
+    MkosiPrinter.info("            FS Compression: " + yes_no(cast(bool, args.compress)) + detail)
```
Comment (behrmann) [inline mkosi/__init__.py:5741]: Why the cast here, if the signature is changed above to `Optional[bool]`?

**Review by behrmann — COMMENTED** [review]
Excellent ground work!

File: mkosi/__init__.py (hunk 1 above)
Comment (DaanDeMeyer) [reply mkosi/__init__.py:5741]: Not sure what you mean, `compress` was changed to `Union[str, bool, None]`.

Comment (behrmann) [reply mkosi/__init__.py:5741]: Sorry, my bad. Guess I wasn't fully awake yet, when I looked over this.

**Comment by DaanDeMeyer** [comment]
@behrmann Any more comments on this? If not, I'll merge

**Comment by behrmann** [comment]
> @behrmann Any more comments on this? If not, I'll merge

Nope, I think this is good to go and everything else can be changed later.

**Pull request merged**